### PR TITLE
Implement proper number parsing

### LINKED ISSUE #101: Hex numbers are valid constants too
```
diff --git a/lib/Guacamole.pm b/lib/Guacamole.pm
index caff3ff..0ef4bcb 100644
--- a/lib/Guacamole.pm
+++ b/lib/Guacamole.pm
@@ -1376,11 +1376,16 @@ VersionNumberSegment ~ [0-9] [0-9] [0-9]
                      | [0-9] [0-9]
                      | [0-9]

+HexPrefix ~ '0x'
+HexDigits ~ [0-9a-fA-F]+
+HexNumber ~ HexPrefix HexDigits
+
 LitNumber ::= LitNumberDec
 LitNumberDec ~ Digits Period Digits
              | Digits Period
              | Period Digits
              | Digits
+             | HexNumber

 Period      ~ '.'
 Digits      ~ [0-9]+
```
It is too silly not too allow hex numeric constants.

### PR DESCRIPTION
**Pull request by vickenty**
Parse decimal, octal, hexadecimal and binary numbers, exponents and visual separators in weird places.

Alternative to, closes #103
Fixes #101

### CODE REVIEW

File: lib/Guacamole.pm [hunk 1]
```diff
@@ -1377,12 +1377,58 @@ VersionNumberSegment ~ [0-9] [0-9] [0-9]
...
+
+NumberDec ~ NumberDecInt
+          | NumberDecInt ExpDec
+          | NumberDecInt '.' DigitsDec
+          | NumberDecInt '.' DigitsDec ExpDec
+          | '.' DigitDec DigitsDec
+          | '.' DigitDec DigitsDec ExpDec
+
+NumberDecInt ~ [1-9] DigitsDec
+             | '0'
```
Comment (gonzus) [inline lib/Guacamole.pm:1397]: The treatment for `Underbars` is missing here, although maybe it is dealt with somewhere else (due to the need to exclude `0` for decimal digits)?

File: t/Statements/Expressions/Value/Literal.t [hunk 2]
```diff
@@ -3,12 +3,77 @@ use warnings;
...
+parses("0.");
+parses(".0");
+
+parses(".0e1");
+parses("0.e1");
+parsent(".e1");
+
+parses("0._4");
+parses(".4_");
+parsent("._4");
```
Comment (gonzus) [inline t/Statements/Expressions/Value/Literal.t:22]: Related to the previous comment, I would add here some test cases for `Underbars` before the decimal dot.

Comment (vickenty) [reply t/Statements/Expressions/Value/Literal.t]: Good idea, thanks.

File: lib/Guacamole.pm (hunk 1 above)
Comment (gonzus) [reply lib/Guacamole.pm:1397]: It just seemed to me that the code for decimal numbers was different from the code for other types of numbers. For hex / oct / bin the rules obviously covered underscores, it wasn't obvious to me that the rules for decimals did too. But I also tried to say that maybe I was just me not seeing it.

I guess, bottom line, if the tests pass for stuff like `1_234_567` then everything is OK.

File: t/Statements/Expressions/Value/Literal.t (hunk 2 above)
Comment (gonzus) [reply t/Statements/Expressions/Value/Literal.t]: By which I meant, very confusingly, tests for stuff such as `1_234_567`, `2_00_000` (or however the Indian thing with crores and lakhs is done), etc.

**Comment by xsawyerx** [comment]
@tux, FYI.

File: lib/Guacamole.pm (hunk 1 above)
Comment (vickenty) [reply lib/Guacamole.pm:1397]: Ah, good point. I have added more tests for these cases.

**Pull request merged**